### PR TITLE
add functionality to pure pursuit to calculate curvature with a trailer

### PR DESCRIPTION
**Pull request by chrpro**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Add functionality to pure pursuit to calculate curvature with a trailer

* **What is the current behavior?** (You can also link to an open issue here)
The pure pursuit does not take into account type of vehicle

* **What is the new behavior (if this is a feature change)?**

Pure pursuit will check if the type is truckstate and call a different function to calculate the  

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

Yes, the RRCar needs to add WayWise/vehicles/truckstate.cpp to CMakeLists.txt in order to compile,
this does not look optimal, but I don't know a work around for the moment

* **Other information**:

### CODE REVIEW

**Comment by chrpro** [comment]
As I say I add the code so Ramana has access,  a discussion i needed for how to move forward, 
in order for the code to be compiled by RCCar the truckstate needs to be visible by the linker, the reason is that I made a dynamic cast of vehiclestate to truck state to check the type of the vehicle, if the vehicle is a truck and has a trailer I call another function for calculating the curvature. 
This can be questionable as RRCar should not need the truckstate so feel free to give a good work around

DO NOT MERGE FOR THE MOMENT

**Comment by r2avula** [comment]
> in order for the code to be compiled by RCCar the truckstate needs to be visible by the linker, the reason is that I made a dynamic cast of vehiclestate to truck state to check the type of the vehicle, if the vehicle is a truck and has a trailer I call another function for calculating the curvature. This can be questionable as RRCar should not need the truckstate so feel free to give a good work around

Instead of the dynamic cast in the purepursuitwaypointfollower class, in my opinion its a cleaner approach to add a private boolean variable mHasTrailer, its public get & set methods, and a public getTrailerAngleRadians to the vehiclestate. This way we can get the info required to calculate curvature with trailer directly from vehiclestate at https://github.com/RISE-Dependable-Transport-Systems/WayWise/blob/8ab34220c21b15458524548df39716c145a8535f/autopilot/purepursuitwaypointfollower.cpp#L95

I think we should also expose the trailer wheelbase, low-speed and high-speed gain parameters. I think WayPointFollowerState is a good place for these: https://github.com/RISE-Dependable-Transport-Systems/WayWise/blob/8ab34220c21b15458524548df39716c145a8535f/autopilot/purepursuitwaypointfollower.h#L22

**Comment by chrpro** [comment]
> > in order for the code to be compiled by RCCar the truckstate needs to be visible by the linker, the reason is that I made a dynamic cast of vehiclestate to truck state to check the type of the vehicle, if the vehicle is a truck and has a trailer I call another function for calculating the curvature. This can be questionable as RRCar should not need the truckstate so feel free to give a good work around
> 
> Instead of the dynamic cast in the purepursuitwaypointfollower class, in my opinion its a cleaner approach to add a private boolean variable mHasTrailer, its public get & set methods, and a public getTrailerAngleRadians to the vehiclestate. This way we can get the info required to calculate curvature with trailer directly from vehiclestate at
> 
> https://github.com/RISE-Dependable-Transport-Systems/WayWise/blob/8ab34220c21b15458524548df39716c145a8535f/autopilot/purepursuitwaypointfollower.cpp#L95
> 
> I think we should also expose the trailer wheelbase, forward and backward gain parameters. I think WayPointFollowerState is a good place for these:
> 
> https://github.com/RISE-Dependable-Transport-Systems/WayWise/blob/8ab34220c21b15458524548df39716c145a8535f/autopilot/purepursuitwaypointfollower.h#L22

Sounds good, please commit and push your suggestions

**Comment by r2avula** [comment]
@chrpro: Can you check if everything works now with Griffin, especially the trailer angle?

**Comment by chrpro** [comment]
> @chrpro: Can you check if everything works now with Griffin, especially the trailer angle?

It does not look so good,  I need to test more, but these changes looks like breaking changes to me. 
Also, we need to test that RCCar works correctly, if we modify the vehicle changes

File: autopilot/purepursuitwaypointfollower.cpp
Comment (rickaha) [inline autopilot/purepursuitwaypointfollower.cpp]: I think we need to move getCurvatureWithTrailer() out of purepursuitwaypointfollower. It looks like getCurvatureWithTrailer() is a version of PurepursuitWaypointFollower::getCurvatureToPointInVehicleFrame(), which we should move to a more general place. I propose we move it to vehiclestate and make it virtual, this way we can move getCurvatureWithTrailer() to the truckstate, and the current implementation of PurepursuitWaypointFollower::getCurvatureToPointInVehicleFrame() to the carstate. It would be something like this

VehicleState::getCurvatureToPointInVehicleFrame() // virtual
TruckState::getCurvatureToPointInVehicleFrame() // override
CarState::getCurvatureToPointInVehicleFrame() // override

This way, we can call the function using only vehiclestate like

```suggestion
double PurepursuitWaypointFollower::getCurvatureToPointInENU(const QPointF &point)
{
    QSharedPointer<VehicleState> vehicleState = isOnVehicle() ? mMovementController->getVehicleState() : mVehicleConnection->getVehicleState();

    return vehicleState->getCurvatureToPointInVehicleFrame(coordinateTransforms::ENUToVehicleFrame(point, vehicleState->getPosition(mPosTypeUsed).getXYZ()));
}
```
All following commits can be removed and we dont clutter purepursuitwaypointfollower with variables connected to a specific vehicle.

File: vehicles/vehiclestate.h
Comment (rickaha) [inline vehicles/vehiclestate.h:1]: I think its a good idea to use a flag, but this should be in the truckstate, because its not general

**Review by rickaha — CHANGES_REQUESTED** [review]
I think we need to refactor, to keep purepursuitwaypointfollower more general

File: vehicles/vehiclestate.h
Comment (rickaha) [inline vehicles/vehiclestate.h:78]: I think its a good idea to add a private boolean variable mHasTrailer with public get & set methods, but it should be in truckstate, given its vehicle specific

File: autopilot/purepursuitwaypointfollower.cpp
Comment (r2avula) [reply autopilot/purepursuitwaypointfollower.cpp]: I agree with this. I think it's also more cleaner this way than my earlier suggestion. I will revert the commits to https://github.com/RISE-Dependable-Transport-Systems/WayWise/pull/103/commits/40c78658bb3bf68edea81b5cdedf26003a61ee5e. They are available [here ](https://github.com/RISE-Dependable-Transport-Systems/WayWise/tree/temp/pure-pursuit-with-a-trailer-var) for reference.

@chrpro: As it was your original work, please incorporate these suggestions or let me know otherwise. It would be useful to bring these changes to the main branch so they can be used in WayWiseR and I would be glad to help if needed.

File: vehicles/truckstate.cpp
Comment (rickaha) [inline vehicles/truckstate.cpp:39]: I think we only need to make some changes to this function and then its ready to merge

Comment (rickaha) [inline vehicles/truckstate.cpp:39]: I think we only need to make some changes to this function and then its ready to merge

**Review by r2avula — DISMISSED** [review]
The changes now look good for me and can be merged!👍

File: vehicles/truckstate.cpp
Comment (r2avula) [reply vehicles/truckstate.cpp:39]: The hardcoded parameters are now configurable via Truckstate.

**Comment by rickaha** [comment]
> The changes now look good for me and can be merged!👍

Definitions of functions should be in *.cpp file, not i *.h

TruckState::getCurvatureWithTrailer() is still not fixed, the algorithm is performed in the wrong coordinatesystem and need to be re-written...

**Comment by r2avula** [comment]
> > The changes now look good for me and can be merged!👍
> 
> Definitions of functions should be in *.cpp file, not i *.h
> 
> TruckState::getCurvatureWithTrailer() is still not fixed, the algorithm is performed in the wrong coordinatesystem and need to be re-written...

@rickaha: Good that you mention about the wrong coordinate-system. I thought it was already tested with Griffin! I will then dismiss my approval review. I guess, by function definition in the .h file you are referring to https://github.com/RISE-Dependable-Transport-Systems/WayWise/blob/c7f6eda3ed05af65963a9cb4fa3d7d3c676e05d2/vehicles/truckstate.h#L38 I can move it the .cpp file. I also saw that you merged https://github.com/RISE-Dependable-Transport-Systems/WayWise/commit/08992f1e77c084860eb7f97d0d7176829499cd7c and https://github.com/RISE-Dependable-Transport-Systems/WayWise/commit/510203d74a21d2409dd950c5a1b7fa768f67eadd into https://github.com/RISE-Dependable-Transport-Systems/WayWise/commit/c7f6eda3ed05af65963a9cb4fa3d7d3c676e05d2. IMO, the cleanup of the truckstate and trailerstate files where empty spaces are removed should be a separate commit so it would be easier to check the real changes in diff.

**Comment by rickaha** [comment]
> > > The changes now look good for me and can be merged!👍
> > 
> > 
> > Definitions of functions should be in *.cpp file, not i *.h
> > TruckState::getCurvatureWithTrailer() is still not fixed, the algorithm is performed in the wrong coordinatesystem and need to be re-written...
> 
> @rickaha: Good that you mention about the wrong coordinate-system. I thought it was already tested with Griffin! I will then dismiss my approval review. I guess, by function definition in the .h file you are referring to
> 
> https://github.com/RISE-Dependable-Transport-Systems/WayWise/blob/c7f6eda3ed05af65963a9cb4fa3d7d3c676e05d2/vehicles/truckstate.h#L38
> 
> I can move it the .cpp file. I also saw that you merged [08992f1](https://github.com/RISE-Dependable-Transport-Systems/WayWise/commit/08992f1e77c084860eb7f97d0d7176829499cd7c) and [510203d](https://github.com/RISE-Dependable-Transport-Systems/WayWise/commit/510203d74a21d2409dd950c5a1b7fa768f67eadd) into [c7f6eda](https://github.com/RISE-Dependable-Transport-Systems/WayWise/commit/c7f6eda3ed05af65963a9cb4fa3d7d3c676e05d2). IMO, the cleanup of the truckstate and trailerstate files where empty spaces are removed should be a separate commit so it would be easier to check the real changes in diff.

There will be a lot of unnecessary commits this way. I squashed the last two commits again

**Review by r2avula — APPROVED** [review]
The [getCurvatureWithTrailer()](https://github.com/RISE-Dependable-Transport-Systems/WayWise/pull/103/commits/c1c69e33fa183eb6338fd6fc148f669f8010df99) is now rewritten to account for the coordinate system change to the vehicle frame. It just needed a cleanup of a few redundant variables. Tested with Truck and Trailer in Carla and it works fine!

**Comment by rickaha** [comment]
> The [getCurvatureWithTrailer()](https://github.com/RISE-Dependable-Transport-Systems/WayWise/pull/103/commits/c1c69e33fa183eb6338fd6fc148f669f8010df99) is now rewritten to account for the coordinate system change to the vehicle frame. It just needed a cleanup of a few redundant variables. Tested with Truck and Trailer in Carla and it works fine!

I'll let @chrpro go through and test this on Griffin

**Comment by chrpro** [comment]
@r2avula I tested with the truck and trailer and it works, I add a small modification in the calculations,
please test again with the simulator and if it works go ahead and merge :)

**Review by rickaha — APPROVED** [review]

**Review by r2avula — APPROVED** [review]
Tested in simulation and works as expected!

**Pull request merged**
